File: py/src/day_02.py
from dataclasses import dataclass

from aocd import get_data


@dataclass
class Game:
    i: int
    r: int
    g: int
    b: int

    @classmethod
    def from_str(cls, s):
        prefix, suffix = s.split(":")
        i = int(prefix.split()[-1])
        games = suffix.split(";")
        max_r = max_g = max_b = 0
        for game in games:
            balls = game.split(",")
            for ball in balls:
                count, color = ball.split()
                count = int(count)
                if color == "red":
                    max_r = max(max_r, count)
                elif color == "green":
                    max_g = max(max_g, count)
                elif color == "blue":
                    max_b = max(max_b, count)
        return cls(i, max_r, max_g, max_b)


def preprocess(data):
    return [Game.from_str(line) for line in data.splitlines()]


def part_1(data):
    return sum(
        game.i for game in data if game.r <= 12 and game.g <= 13 and game.b <= 14
    )


def part_2(data):
    return sum(game.r * game.g * game.b for game in data)


test = """\
Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green
"""


def main():
    data = preprocess(get_data(day=2, year=2023))
    print(part_1(preprocess(test)))
    print(part_1(data))
    print(part_2(preprocess(test)))
    print(part_2(data))


if __name__ == "__main__":
    main()

File: rs/src/day_02.rs
pub struct Game {
    i: usize,
    r: usize,
    g: usize,
    b: usize,
}

impl From<&str> for Game {
    fn from(input: &str) -> Self {
        let (prefix, suffix) = input.split_once(":").unwrap();
        let i = prefix.split_once(" ").unwrap().1.parse::<usize>().unwrap();
        let mut max_r = 0;
        let mut max_g = 0;
        let mut max_b = 0;
        for game in suffix.split(";") {
            for ball in game.split(",") {
                let (count, color) = ball.trim().split_once(" ").unwrap();
                match color {
                    "red" => {
                        max_r = max_r.max(count.parse::<usize>().unwrap());
                    }
                    "green" => {
                        max_g = max_g.max(count.parse::<usize>().unwrap());
                    }
                    "blue" => {
                        max_b = max_b.max(count.parse::<usize>().unwrap());
                    }
                    _ => panic!("Unknown color {}", color),
                }
            }
        }
        Self {
            i,
            r: max_r,
            g: max_g,
            b: max_b,
        }
    }
}

#[aoc_generator(day2)]
pub fn get_input(input: &str) -> Vec<Game> {
    input.lines().map(|l| l.into()).collect()
}

#[aoc(day2, part1)]
pub fn part_1(games: &Vec<Game>) -> usize {
    games
        .iter()
        .filter_map(|g| {
            if g.r <= 12 && g.g <= 13 && g.b <= 14 {
                Some(g.i)
            } else {
                None
            }
        })
        .sum()
}

#[aoc(day2, part2)]
pub fn part_2(games: &Vec<Game>) -> usize {
    games.iter().map(|g| g.r * g.g * g.b).sum()
}

#[cfg(test)]
mod tests {
    use super::*;
    const INPUT: &str = include_str!("../test_data/day_02.txt");

    #[test]
    fn test_part_1() {
        assert_eq!(part_1(&get_input(INPUT)), 8);
    }

    #[test]
    fn test_part_2() {
        assert_eq!(part_2(&get_input(INPUT)), 2286);
    }
}
